core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeJoin.java

```diff
@@ -222,6 +277,30 @@ public static boolean isMergeJoinSupported(JoinRelType joinType) {
     return mapping;
   }
 
+  private RelCollation extendCollation(RelCollation collation, List<Integer> keys) {
+    List<RelFieldCollation> fieldsForNewCollation = new ArrayList<>(keys.size());
+    fieldsForNewCollation.addAll(collation.getFieldCollations());
+    Set<Integer> keySet = new HashSet<>(keys);
```

Comment:
Do we need to keep the keys order?

Reply:
No. Only ordering on collations defined on key need to preserve. For example, required collation [foo.a, foo.b, foo.c ,foo.d] defined on `foo.a=bar.a and foo.b=bar.b`, either pass down [foo.a, foo.b, foo.c, foo.d] or pass down [foo.a, foo.b, foo.d, foo.c] will give correct answer (but [a, b]'s ordering must remain).

It is because, for MergeJoin implementation, the pointers moves when left tuple compares to right tuple, and the only requirement is next tuple should be bigger than previous one.

Still use example above:
left join input could be sorted by [a, b, d, c], and right join input is sorted by [a, b]. So 
1) `left_tuple.compare(right_tuple) < 0`, move left pointer, of course next tuple will be both bigger than a) the previous left tuple b) the current right tuple (due to they are all sorted by [a, b]).
2) same for ``left_tuple.compare(right_tuple) > 0` case.

Reply:
to add a bit explanation on the example above, the only usefulness of pushed collocations on [d, c], is just to maintain increasing order when move left pointer on left join input during merging. Thus [c, d] will achieve the same thing.

Reply:
Why not just use bitset.except to compute the diff keys?

Reply:
Will check how to use `bitset.except`. 

Meanwhile, in fact, the ordering does matter for EnumerableMergeJoin, it is because key selector and key comparator are constructed by sort key ordering.

Thus we have to follow sort key ordering to extend required collations.

Reply:
Use `bitset.except` now to compute diff.